### PR TITLE
Add inclusions/exclusions logic

### PR DESCRIPTION
**Pull request by rita-sonarsource**
Fixes [#3061](https://github.com/SonarSource/sonarlint-visualstudio/issues/3061)

Code coverage is low due to an empty class that would be implemented in a separate task.

### CODE REVIEW

File: src/Core/Analysis/IExclusionSettingsFileStorage.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.Core.Analysis
+{
+    public interface IExclusionSettingsFileStorage
```
Comment (rita-sonarsource) [inline src/Core/Analysis/IExclusionSettingsFileStorage.cs:25]: Based on Ugras' PR

File: src/Integration/Exclusions/ExclusionSettingsFileStorage.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+using System.ComponentModel.Composition;
+using SonarLint.VisualStudio.Core.Analysis;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.Integration.Exclusions
+{
+    [Export(typeof(IExclusionSettingsFileStorage))]
+    internal class ExclusionSettingsFileStorage : IExclusionSettingsFileStorage
```
Comment (rita-sonarsource) [inline src/Integration/Exclusions/ExclusionSettingsFileStorage.cs:28]: Just dummy implementation for now

File: src/Integration/Exclusions/AnalyzableFileIndicator.cs [hunk 3]
```diff
@@ -27,9 +29,49 @@ namespace SonarLint.VisualStudio.Integration.Exclusions
...
+
+            bool IsExcludedInProjectOrGlobally() =>
+                IsExcluded(serverExclusions.Exclusions) ||
+                IsExcluded(serverExclusions.GlobalExclusions);
+
+            bool IsExcluded(string[] exclusions) =>
+                exclusions != null &&
+                exclusions.Any(IsMatch);
+
+            bool IsMatch(string pattern) => globPatternMatcher.IsMatch(pattern, filePath);
```
Comment (rita-sonarsource) [inline src/Integration/Exclusions/AnalyzableFileIndicator.cs:74]: At the moment I'm comparing the pattern as-is to the file path, disregarding the root directory from which we should start. I added a separate task to investigate this.

File: src/Integration/Exclusions/AnalyzableFileIndicator.cs [hunk 4]
```diff
@@ -27,9 +29,49 @@ namespace SonarLint.VisualStudio.Integration.Exclusions
...
+                return true;
+            }
+
+            var shouldAnalyze = IsIncluded() && !IsExcludedInProjectOrGlobally();
+
+            return shouldAnalyze;
+
+            bool IsIncluded() =>
+                serverExclusions.Inclusions == null ||
+                serverExclusions.Inclusions.Length == 0 ||
```
[the remark this answers is not in the document]
Comment (rita-sonarsource) [reply src/Integration/Exclusions/AnalyzableFileIndicator.cs:63]: For inclusions, if the array is empty, the item is considered included. Whereas if the array is not empty, then the item might be excluded.
In exclusions, an empty array and a non-empty array behave the same. Hence why there is no need to check for empty.

File: src/Integration/Exclusions/AnalyzableFileIndicator.cs [hunk 5]
```diff
@@ -27,9 +29,49 @@ namespace SonarLint.VisualStudio.Integration.Exclusions
...
+                serverExclusions.Inclusions.Any(IsMatch);
+
+            bool IsExcludedInProjectOrGlobally() =>
```
Comment (duncanp-sonar) [inline src/Integration/Exclusions/AnalyzableFileIndicator.cs:66]: Do the nested functions improve the readability?

Comment (rita-sonarsource) [reply src/Integration/Exclusions/AnalyzableFileIndicator.cs:66]: You tell me :D

File: src/Integration.UnitTests/Exclusions/AnalyzableFileIndicatorTests.cs [hunk 6]
```diff
@@ -18,19 +18,198 @@
...
+
+            var testSubject = CreateTestSubject(exclusionConfig, patternMatcher.Object);
+
+            var result = testSubject.ShouldAnalyze(TestedFilePath);
+
+            result.Should().Be(expectedResult);
+        }
+
+        [TestMethod]
+
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Exclusions/AnalyzableFileIndicatorTests.cs:147]: FYI extra blank line in after the attribute in this test and the next one.

File: src/Integration/Exclusions/AnalyzableFileIndicator.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/Integration/Exclusions/AnalyzableFileIndicator.cs:63]: Ok, makes sense, but it's not obvious. An explicit comment would help.

File: src/Core/Analysis/IExclusionSettingsFileStorage.cs (hunk 1 above)
Comment (duncanp-sonar) [reply src/Core/Analysis/IExclusionSettingsFileStorage.cs:25]: @ugras-ergun-sonarsource do you mean you'll push the change to 'branch-exclusions`, or you'll create a PR?

Comment (ugras-ergun-sonarsource) [reply src/Core/Analysis/IExclusionSettingsFileStorage.cs:25]: I have a pr already. I had made some changes to signatures locally but I've reverted it back and in the confusion, I have created noise you can disregard my comment.

**Pull request merged**